edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsWorker.cs

```diff
@@ -0,0 +1,129 @@
+                List<Metric> metricsToPersist = new List<Metric>();
+
+                foreach (var moduleResult in await this.scraper.ScrapeAsync(cancellationToken))
+                {
+                    IEnumerable<Metric> scrapedMetrics = PrometheousMetricsParser.ParseMetrics(this.systemTime.UtcNow, moduleResult.Value);
+
+                    foreach (Metric scrapedMetric in scrapedMetrics)
+                    {
+                        // Get the previous scrape for this metric
+                        if (this.metrics.TryGetValue(scrapedMetric.HashNameAndTag(), out Metric oldMetric))
+                        {
+                            // If the metric is unchanged, do nothing
+                            if (oldMetric.Value.Equals(scrapedMetric.Value))
+                            {
+                                continue;
+                            }
+
+                            // if the metric has changed, write the previous metric to disk
+                            metricsToPersist.Add(oldMetric);
+                        }
+
+                        // if new metric or metric changed, save to local buffer
+                        this.metrics[scrapedMetric.HashNameAndTag()] = scrapedMetric;
+                    }
+                }
+
+                Log.LogInformation("Scraped Metrics");
+
+                if (metricsToPersist.Count != 0)
+                {
+                    Log.LogInformation("Storing Metrics");
+                    this.storage.WriteData(Newtonsoft.Json.JsonConvert.SerializeObject(metricsToPersist));
+                    Log.LogInformation("Stored Metrics");
+                }
+            }
+        }
+
+        async Task Upload(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Uploading Metrics");
+                await this.uploader.UploadAsync(this.GetMetricsToUpload(this.lastUploadTime), cancellationToken);
+                this.lastUploadTime = this.systemTime.UtcNow;
+                Log.LogInformation("Uploaded Metrics");
+            }
+        }
+
+        IEnumerable<Metric> GetMetricsToUpload(DateTime lastUploadTime)
+        {
+            // Get all metrics that have been stored since the last upload
+            foreach (KeyValuePair<DateTime, Func<string>> data in this.storage.GetData(lastUploadTime))
+            {
+                var fileMetrics = Newtonsoft.Json.JsonConvert.DeserializeObject<Metric[]>(data.Value()) ?? Enumerable.Empty<Metric>();
+                foreach (Metric metric in fileMetrics)
+                {
+                    yield return metric;
+                }
+            }
+
+            // Get all metrics stored in the local buffer.
+            foreach (Metric metric in this.metrics.Values)
+            {
+                yield return metric;
+            }
+
+            this.storage.RemoveOldEntries(lastUploadTime);
```

Comment:
I don't think it is appropriate to call RemoveOldEntries method in this method; this should be handled by caller.